# 🐍 Python 103 — Beginner Fundamentals
Welcome to Python 103! This guide will help you understand Python from the ground up. No prior programming experience required.

To ensure a deeper understanding and proper logic processing, please complete the 102 notebook (prerequisite) first.

# Pandas: Getting Started

*Pandas* - This is a Python library used for working with datasets.

- It does by helping us make sense of *DataFrames*

*DataFrames* - A form of two-dimensional structured data, like a table with columns and rows. 

## Importing Data | CSV Files

*CSV* - Common Separated Values, and it's a file type that allows data to be saved in a table.

* Data presented in a tabular format is called *structured data*

In [ ]:
# Using Dictionaries to create a dataset of students and their grades

import pandas as pd

# Create a dictionary with student names as keys and their grades as values
student_grades = {
    "Alice": 85,
    "Bob": 92,
    "Charlie": 78,
    "David": 90,
    "Eve": 88
}   
# Convert the dictionary to a pandas DataFrame
pd.DataFrame.from_dict(student_grades, orient='index',columns=['Grade'])

# # Create a DataFrame from the dictionary
# df = pd.DataFrame(list(student_grades.items()), columns=['Student', 'Grade'])

# # Display the DataFrame
# print(df)

,Grade
Alice,85
Bob,92
Charlie,78
David,90
Eve,88


## Using JSON library 

In [ ]:
# Using JSON library to parse a JSON string into a Python dictionary

import json

info = """ {
    "firstName": "Alice",
    "lastName": "Smith",
    "age": 30,
    "city": "Stockholm",
    "hobbies": ["reading", "traveling", "cooking"]
}"""

# Parse JSON string into Python dictionary
info_dict = json.loads(info)

info_dict

{'firstName': 'Alice',
 'lastName': 'Smith',
 'age': 30,
 'city': 'Stockholm',
 'hobbies': ['reading', 'traveling', 'cooking']}

In [10]:
info_dict["city"]

'Stockholm'

## Loading data from Compressed Files

In big data ere - it is likely to read data from compressed files.

One way to unzip the data is to use gzip. 

In [11]:
import gzip 
import json

with gzip.open('data/poland-bankruptcy-data-2008.json.gz', 'rt') as f:
    poland_data = json.load(f)

In [17]:
# Check the type of poland_data
# type(poland_data)

# Check the keys in the poland_data dictionary
poland_data.keys()

dict_keys(['schema', 'data', 'metadata'])

In [ ]:
# Using the `from_dict`  Pandas function to read the data

# DataFrame()
df = pd.DataFrame.from_dict(poland_data['data'])

df.head()

,company_id,feat_1,feat_2,feat_3,feat_4,feat_5,feat_6,feat_7,feat_8,feat_9,...,feat_56,feat_57,feat_58,feat_59,feat_60,feat_61,feat_62,feat_63,feat_64,bankrupt
0,1,0.202350,0.46500,0.240380,1.5171,-14.547,0.510690,0.25366,0.91816,1.15190,...,0.13184,0.473950,0.86816,0.00024,8.5487,5.16550,107.740,3.38790,5.3440,False
1,2,0.030073,0.59563,0.186680,1.3382,-37.859,-0.000319,0.04167,0.67890,0.32356,...,0.12146,0.074369,0.87235,0.00000,1.5264,0.63305,622.660,0.58619,1.2381,False
2,3,0.257860,0.29949,0.665190,3.2211,71.799,0.000000,0.31877,2.33200,1.67620,...,0.16499,0.369210,0.81614,0.00000,4.3325,3.19850,65.215,5.59690,47.4660,False
3,4,0.227160,0.67850,0.042784,1.0828,-88.212,0.000000,0.28505,0.47384,1.32410,...,0.29358,0.706570,0.78617,0.48456,5.2309,5.06750,142.460,2.56210,3.0066,False
4,5,0.085443,0.38039,0.359230,1.9444,21.731,0.187900,0.10823,1.37140,1.11260,...,0.10124,0.163790,0.89876,0.00000,5.7035,4.00200,89.058,4.09840,5.9874,False


# Pickle File
A pickle file is a file that stores a Python object in a serialized binary format using the built-in Python module called `pickle`

Understand Pickling - Serialization and Deserializing.

`Serialization` means converting an object (like a list, dictionary, machine learning model, etc.) into bytes so it can be:
* saved to disk,
* transferred over a network,
* or loaded later.

A pickle file commonly uses extensions like:

* .pkl
* .pickle

`Deserializing` is the reverse process, turning a stream of bytes into an object in memory.

In [ ]:
# Example implementation using a dictionary with lists as values
fruits = {
    "apple": ["red", "green"],
    "banana": ["yellow", "green"],
    "grape": ["purple", "green"],
    "watermelon": ["green", "red"],
    "peach": ["pink", "yellow"],
    "pear": ["green", "yellow"],
    "pawpaw": ["green", "yellow"]
    
}

fruits

{'apple': ['red', 'green'],
 'banana': ['yellow', 'green'],
 'grape': ['purple', 'green'],
 'watermelon': ['green', 'red'],
 'peach': ['pink', 'yellow'],
 'pear': ['green', 'yellow'],
 'pawpaw': ['green', 'yellow']}

In [6]:
import pickle 
import os


'''
"wb" means write binary
pickle.dump() writes the object into the file
'''

# create folder if it doesn't exist
os.makedirs('./pickle-files', exist_ok=True)

pickle.dump(fruits, open('./pickle-files/fruits.pkl', 'wb'))



The `pickle-files` inside this directory `notebooks/python-basics/pickle-files` contains a file named `fruits.pkl`

In [9]:
# We can read the pickled file back into a Python object using pickle.load()

with open('./pickle-files/fruits.pkl', "rb") as f:
    unpickled_fruits = pickle.load(f)

# Call the unpickled_fruits variable to see the contents
unpickled_fruits

{'apple': ['red', 'green'],
 'banana': ['yellow', 'green'],
 'grape': ['purple', 'green'],
 'watermelon': ['green', 'red'],
 'peach': ['pink', 'yellow'],
 'pear': ['green', 'yellow'],
 'pawpaw': ['green', 'yellow']}

# DataFrame

## Step 1: Import Pandas alias `pd`
Pandas is a library...

In [21]:
import pandas as pd

df = pd.read_csv('../../data/colombia-real-estate-1.csv')
df.head()

,property_type,department,lat,lon,area_m2,price_usd
0,house,Bogotá D.C,4.690,-74.048,187.0,"$330,899.98"
1,house,Bogotá D.C,4.695,-74.082,82.0,"$121,555.09"
2,house,Quindío,4.535,-75.676,235.0,"$219,474.47"
3,house,Bogotá D.C,4.620,-74.129,195.0,"$97,919.38"
4,house,Atlántico,11.012,-74.834,112.0,"$115,477.34"


## Step 2: DataFrame Indices

A DataFrame stores data in a row-and-column format 
* DataFrame index is a special kind or column that helps identify the location of each row.
* The default index uses integers starting at zero, but one can also setup customized indices like `name`, `location`

In [ ]:
# Example 

df = pd.read_csv('../../data/colombia-real-estate-1.csv')

df.head(5)

,property_type,department,lat,lon,area_m2,price_cop
0,house,Magdalena,11.233,-74.204,235.0,4.000000e+08
1,house,Bogotá D.C,4.714,-74.030,130.0,8.500000e+08
2,house,Cundinamarca,4.851,-74.059,137.0,4.750000e+08
3,house,Atlántico,11.006,-74.808,346.0,1.400000e+09
4,house,Cundinamarca,4.857,-74.061,175.0,4.300000e+08


In [45]:
# We can call the index column through `.index` attribute
df.index[0:5]

# --- IGNORE --- Alternative way to call the index column
# df.index[:5]

RangeIndex(start=0, stop=5, step=1)

Using the `set_index` method, one can set the column `department` as the index instead.
`Note`
* Index column cannot have duplicate rows, example: we cannot set `property_type` as the index column

In [46]:
df.set_index("department", inplace=True)

df.head(5)

,property_type,lat,lon,area_m2,price_cop
department,,,,,
Magdalena,house,11.233,-74.204,235.0,4.000000e+08
Bogotá D.C,house,4.714,-74.030,130.0,8.500000e+08
Cundinamarca,house,4.851,-74.059,137.0,4.750000e+08
Atlántico,house,11.006,-74.808,346.0,1.400000e+09
Cundinamarca,house,4.857,-74.061,175.0,4.300000e+08



`Note` how the index column is now the department column


Using the `reset_index()` function, one can reset the index back to default integer counts (id), and `department` will become a column again

In [48]:
df.reset_index(inplace=True)

df.head()

,index,department,property_type,lat,lon,area_m2,price_cop
0,0,Magdalena,house,11.233,-74.204,235.0,4.000000e+08
1,1,Bogotá D.C,house,4.714,-74.030,130.0,8.500000e+08
2,2,Cundinamarca,house,4.851,-74.059,137.0,4.750000e+08
3,3,Atlántico,house,11.006,-74.808,346.0,1.400000e+09
4,4,Cundinamarca,house,4.857,-74.061,175.0,4.300000e+08


## Step 3: Inspecting DataFrames

This involves inspecting a created DataFrame in order to see what's there. 

Examples of ways to inspect a DataFrame include:
* shape 
* info
* head

If interested in understanding `dimensionality` of a DataFrame:

* we can use the `.shape` attribute to get the number of rows and columns in the DataFrame. 

The output will be a tuple where: 
* the first element is the number of rows 
* and the second element is the number of columns.   

In [49]:
df.shape

(3066, 7)

The output shows that the `colombia-real-estate-1` DataFrame has 
* `3066 rows` and `6 columns`

To get a `general idea` of what the DataFrame: 
* one can use `info` method

In [50]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3066 entries, 0 to 3065
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          3066 non-null   int64  
 1   department     3066 non-null   str    
 2   property_type  3066 non-null   str    
 3   lat            2958 non-null   float64
 4   lon            2958 non-null   float64
 5   area_m2        3066 non-null   float64
 6   price_cop      3066 non-null   float64
dtypes: float64(4), int64(1), str(2)
memory usage: 216.4 KB


`info` output provides details about the column such as:
* the number of columns
* the names of the columns
* the data type for each column 
* how many non-null rows are contained in the DataFrame

Returns a specified number of rows, string from the top
* `head()` method returns the first 5 rows if a number is not specified.

In [52]:
df.head(7)

# --- IGNORE ---
# print(df.head(7))

,index,department,property_type,lat,lon,area_m2,price_cop
0,0,Magdalena,house,11.233,-74.204,235.0,4.000000e+08
1,1,Bogotá D.C,house,4.714,-74.030,130.0,8.500000e+08
2,2,Cundinamarca,house,4.851,-74.059,137.0,4.750000e+08
3,3,Atlántico,house,11.006,-74.808,346.0,1.400000e+09
4,4,Cundinamarca,house,4.857,-74.061,175.0,4.300000e+08
5,5,Cundinamarca,house,4.883,-74.035,297.0,1.500000e+09
6,6,Atlántico,house,11.020,-74.864,170.0,5.500000e+08


### Sorting 
DataFrame allows for sorting of data.

Pandas provides two sorting methods,
* sort_values
* sort_index

In [53]:
df.head(10)

,index,department,property_type,lat,lon,area_m2,price_cop
0,0,Magdalena,house,11.233,-74.204,235.0,4.000000e+08
1,1,Bogotá D.C,house,4.714,-74.030,130.0,8.500000e+08
2,2,Cundinamarca,house,4.851,-74.059,137.0,4.750000e+08
3,3,Atlántico,house,11.006,-74.808,346.0,1.400000e+09
4,4,Cundinamarca,house,4.857,-74.061,175.0,4.300000e+08
5,5,Cundinamarca,house,4.883,-74.035,297.0,1.500000e+09
6,6,Atlántico,house,11.020,-74.864,170.0,5.500000e+08
7,7,Bogotá D.C,house,4.730,-74.066,65.0,2.500000e+08
8,8,Valle del Cauca,house,3.328,-76.639,105.0,3.400000e+08
9,9,Antioquia,house,6.189,-75.563,171.0,7.700000e+08


We can sort the whole DataFrame by values of a column

In [63]:
df.set_index("price_cop").sort_index(ascending=False).head(10)

,index,department,property_type,lat,lon,area_m2
price_cop,,,,,,
1.700000e+09,2558,Bogotá D.C,house,4.703,-74.046,323.0
1.700000e+09,1614,Bogotá D.C,house,4.663,-74.047,155.0
1.700000e+09,2045,Cundinamarca,house,4.854,-74.072,445.0
1.700000e+09,37,Bogotá D.C,house,4.736,-74.025,196.0
1.700000e+09,2329,Bogotá D.C,house,4.724,-74.076,247.0
1.700000e+09,2207,Bogotá D.C,house,4.709,-74.028,257.0
1.700000e+09,464,Cundinamarca,house,4.850,-74.079,445.0
1.700000e+09,694,Cundinamarca,house,4.868,-74.037,250.0
1.700000e+09,1366,Cundinamarca,house,4.585,-74.188,380.0


In [60]:
df.sort_values("department").head(20)

,index,department,property_type,lat,lon,area_m2,price_cop
9,9,Antioquia,house,6.189000,-75.563000,171.0,7.700000e+08
10,10,Antioquia,house,6.238000,-75.569000,161.0,2.049000e+08
22,22,Antioquia,house,6.164000,-75.616000,260.0,5.200000e+08
55,55,Antioquia,house,6.122121,-75.419769,270.0,6.200000e+08
122,122,Antioquia,house,6.212000,-75.556000,135.0,6.285980e+08
125,125,Antioquia,house,6.462000,-75.556000,220.0,4.800000e+08
133,133,Antioquia,apartment,6.242976,-75.601785,152.0,4.300000e+08
190,190,Antioquia,house,NaN,NaN,156.0,5.000000e+08
209,209,Antioquia,house,6.152502,-75.560532,300.0,8.800000e+08
264,264,Antioquia,house,NaN,NaN,108.0,3.250000e+08


### Working with Columns